webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that "a STOP_SENDING frame requests that the receiving endpoint send a
+RESET_STREAM frame.". See [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5).
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+## Connection Security
+
+Note that the below uses the message framing described in
+[multiplexing](#multiplexing).
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel.
+
+2. _A_ starts a Noise `XX` handshake using _A_'s and _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+   Instead of exchanging the TLS certificate fingerprints on the established
+   Noise channel once the Noise handshake succeeded, _A_ and _B_ use the [Noise
+   Prologue](https://noiseprotocol.org/noise.html#prologue) mechanism, thus
+   saving one round trip.
+
+   More specifically _A_ and _B_ set the Noise _Prologue_ to
+   `libp2p-webrtc-noise:<FINGERPRINTS>` before starting the actual Noise
+   handshake. `<FINGERPRINTS>` is the concatenation of the two TLS
+   fingerprints of _A_ and _B_ in their multihash byte representation, sorted in
+   ascending order.
```

Comment:
I think it would be useful to provide test vector for this as part of the specification. Those could easily be turned into unit tests for each implementation.

Reply:
Good idea. @thomaseizinger would you mind proposing a pull request against this pull request adding one or more examples?

Reply:
I'll see to add some.

Reply:
I have a first iteration of such tests: [libp2p/rust-libp2p@`c709da9` (#2622)](https://github.com/libp2p/rust-libp2p/pull/2622/commits/c709da95df7b222c6ea79f3d1404b478f3fcbf9c)

What do you think?

Reply:
Opened PR here: https://github.com/mxinden/specs/pull/2

Reply:
Resolving here. Added the above pull request to my TODO list to follow up with.